### PR TITLE
RichText: don't set focus when applying format

### LINKED ISSUE #17505: Changing a RichText value automatically focus the RichText
**Describe the bug**
Currently, if we change a RichText value the RichText gets focused and becomes the active element event if another element was focused and rich text value is changed using a programmatic interface e.g: `wp.data.dispatch( 'core/block-editor' ).updateBlockAttributes( block.clientId, { content: '1<strong>234</strong>56789' } );`.

Previously the popover was closed when a click outside them happened; now they are closed when the focus is outside them. This change makes the rich text automatically gaining focus problem worse. For example, if we have popover with a SelectControl that allows switching the content of a paragraph between some predefined values when the select is used, the popover closes.

Another side effect is that on PR https://github.com/WordPress/gutenberg/pull/16014 that adds custom color as a format, each time we try to use the color picker the popover closes, because when we use the color picker there is a change in the format leaving to a change to the content of paragraph and making the focus go outside the popover.
I guess we should address this issue before WordPress 5.3 because this problem may affect third-party plugins that add color options or even other option in a popover.

I guess we may also have a11y related problems with moving the focus.

**To reproduce**
- Insert paragraph block:
```
const block = wp.blocks.createBlock( 'core/paragraph', { content: '123456789' } );
wp.data.dispatch( 'core/block-editor' ).insertBlock( block );
```
- Put the cursor inside the paragraph e.g. in the middle of the paragraph.
Focus some element on the page (and verify it gets focused):
```
document.querySelector( 'button.editor-block-navigation' ).focus();
console.log( document.activeElement );
````
- Programatically change the paragraph content and verify the paragraph becomes the active element:
```
	wp.data.dispatch( 'core/block-editor' ).updateBlockAttributes( block.clientId, { content: '1<strong>234</strong>56789' } );
	console.log( document.activeElement );
```

Sep-20-2019 18-50-38

cc: @ellatrix

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #17505.

Replacement for #17617. This PR separates focus and applying format so that focus isn't immediately set back while formatting is applied multiple times.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by ellatrix** [comment]
@jorgefilipecosta Does this work for you?

File: packages/block-editor/src/components/rich-text/index.js [hunk 1]
```diff
@@ -411,9 +411,9 @@ class RichTextWrapper extends Component {
 				start={ start }
 				reversed={ reversed }
 			>
-				{ ( { isSelected, value, onChange, Editable } ) =>
+				{ ( { isSelected, value, onChange, onFocus, Editable } ) =>
```
Comment (jorgefilipecosta) [inline packages/block-editor/src/components/rich-text/index.js]: I'm not sure about the name onFocus, it indicates that it is an event e.g: that we can pass something that is called when something gains focus, while what happens is that onFocus is a function that excitedly puts that focus back on rich text. Maybe we can call it "focus" or "focusRichText"?

File: packages/format-library/src/bold/index.js [hunk 2]
```diff
@@ -13,8 +13,15 @@ export const bold = {
...
-	edit( { isActive, value, onChange } ) {
-		const onToggle = () => onChange( toggleFormat( value, { type: name } ) );
+	edit( { isActive, value, onChange, onFocus } ) {
+		function onToggle() {
+			onChange( toggleFormat( value, { type: name } ) );
+		}
+
+		function onClick() {
+			onToggle();
+			onFocus();
```
Comment (jorgefilipecosta) [inline packages/format-library/src/bold/index.js]: This change is something that will impact third-party developers creating their own formats, I guess we should refer this PR in a dev note.

**Review by jorgefilipecosta — APPROVED** [review]
I like that the focus back to rich text is something explicitly that developers of custom formats can control if/when it happens. I think this API is more flexible.
Everything worked well on my tests, https://github.com/WordPress/gutenberg/issues/17505 is fixed and https://github.com/WordPress/gutenberg/pull/16014 is unblocked with this PR. Nice work 👍

File: packages/block-editor/src/components/rich-text/index.js (hunk 1 above)
Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js]: What's the difference with `onChange`? It's also not an event handlers but rather an action to change something. I'd prefer to keep it consistent with `onChange`, which is maybe also named poorly?

File: packages/format-library/src/bold/index.js (hunk 2 above)
Comment (ellatrix) [reply packages/format-library/src/bold/index.js]: Added the label. I won't break anything, but yes it's a change in behaviour. :)

File: packages/block-editor/src/components/rich-text/index.js (hunk 1 above)
Comment (ellatrix) [reply packages/block-editor/src/components/rich-text/index.js]: Cc @aduth since I know you have opinions about this. :)

**Comment by mcsf** [comment]
Noticed this bug on this branch when attempting to unlink by tabbing backwards:

keyboard-focus-unlink

**Review by getdave — COMMENTED** [review]
@ellatrix I tested this against https://github.com/WordPress/gutenberg/pull/19413 [and it now works](https://github.com/WordPress/gutenberg/pull/19413#issuecomment-573112640). Thank you.

**Comment by ellatrix** [comment]
@mcsf I can't reproduce. Tabbing backwards seems to always correctly unlink. Could you provide more detailed steps to reproduce?

**Comment by mcsf** [comment]
> Tabbing backwards seems to always correctly unlink. Could you provide more detailed steps to reproduce?

Seems like I can no longer reproduce either. 🤷‍♀️ Don't let me hold this PR back!

**Comment by ellatrix** [comment]
@mcsf Thanks for testing and @jorgefilipecosta thanks for reviewing!

**Pull request merged**